auth_mellon_util.c

```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
     return url;
 }
 
-/* Thus function checks if an HTTP PAOS header is valid.
+/*------------------------ Begin Token Parsing Code --------------------------*/
+
+typedef enum {
+    TOKEN_WHITESPACE = 1,
+    TOKEN_SEMICOLON,
+    TOKEN_COMMA,
+    TOKEN_EQUAL,
+    TOKEN_IDENTIFIER,
+    TOKEN_DBL_QUOTE_STRING,
+} TokenType;
+
+typedef struct {
+    TokenType type;             /* The type of this token */
+    const char *str;            /* The string value of the token */
+    apr_size_t len;             /* The number of characters in the token */
+    apr_size_t offset;          /* The offset from the beginning of
+                                   the string to the start of the token */
+} Token;
+
+
+#ifdef DEBUG
+/* Return string representation of TokenType enumeration
+ *
+ * Parameters:
+ *  token_type  A TokenType enumeration
+ * Returns:     String name of token_type
+ */
+static const char *
+token_type_str(TokenType token_type)
+{
+    switch(token_type) {
+    case TOKEN_WHITESPACE:       return "WHITESPACE";
+    case TOKEN_SEMICOLON:        return "SEMICOLON";
+    case TOKEN_COMMA:            return "COMMA";
+    case TOKEN_EQUAL:            return "EQUAL";
+    case TOKEN_IDENTIFIER:       return "IDENTIFIER";
+    case TOKEN_DBL_QUOTE_STRING: return "DBL_QUOTE_STRING";
+    default:                     return "unknown";
+    }
+}
+
+static void dump_tokens(request_rec *r, apr_array_header_t *tokens)
+{
+    apr_size_t i;
+    
+    for (i = 0; i < tokens->nelts; i++) {
+        Token token = APR_ARRAY_IDX(tokens, i, Token);
+        ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r,
+                      "token[%2zd] %s \"%s\" offset=%lu len=%lu ", i,
+                      token_type_str(token.type), token.str,
+                      token.offset, token.len);
+    }
+}
+#endif
+
+/* Break a string into a series of tokens
  *
- * A PAOS header must be composed of 2 values seperated by a semicolon.
+ * Given a string return an array of tokens. If the string cannot be
+ * successfully parsed an error string is returned at the location
+ * specified by the error parameter, if error is NULL then the parsing
+ * was successful. If an error occured the returned array of tokens
+ * will include all tokens parsed up until where the unrecognized
+ * input occurred. The input str is never modified.
  *
- * The first value must be a version declaration in the form ver="xxx"
- * (note the version string must be in double quotes).
+ * Parameters:
+ *  pool              memory allocation pool
+ *  str               input string to be parsed.
+ *  ignore_whitespace if True whitespace tokens are not returned
+ *  error             location where error string is returned
+ *                    if NULL no error occurred
+ * Returns:           array of Token objects
+ */
+
+#define TOKEN_INIT(t) memset(&t, 0, sizeof(t))
+
+#define PUSH_TOKEN(token_type, start, end)              \
```

Comment:
Is one allowed to access `tokens->pool`? In that case, I would suggest replacing this `#define` with a `static inline` function:

```
static inline void push_token(apr_array_header_t *tokens, token_type, start, end)
```

Using a `#define` makes the code hard to read, especially as you try to match variables.